0005-ordered-quantification.md

```diff
@@ -0,0 +1,661 @@
+Any other `decreases` clause can be provided
+to ensure the loop terminates even if the domain is potentially infinite. For example, the following example collects
+at most five arbitrary primes (assuming that Dafny can figure out how to enumerate the `allPrimes` infinite set):
+
+```dafny
+var allPrimes := iset n | !exists i, j | 1 < i <= j < n :: i * j == n;
+var fivePrimes := {};
+foreach p <- allPrimes
+  invariant |fivePrimes| <= 5
+  decreases 5 - |fivePrimes|
+{
+  if |fivePrimes| == 5 { break; }
+  fivePrimes := fivePrimes + [x];
+}
+// We can't assert |fivePrimes| == 5 here, since
+// we haven't proven that the loop couldn't have terminated
+// earlier by running out of primes.
+// The decreases metric proves that the loop will terminate,
+// but is only an upper bound on the number of iterations.
+assert |fivePrimes| <= 5;
+```
+
+There will not initially (see Open Questions) be any builtin support for automatically tracking the enumerated values so far, 
+as there is for `iterator` values (see https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-iterator-types). 
+It is straightforward to track these values manually, however:
+
+```dafny
+ghost var xs := [];
+foreach x <- s {
+  ...
+  xs := xs + [x];
+}
+```
+
+It is even possible to attach this ghost state directly to sequences with a helper function:
+
+```dafny
+foreach xAndXs <- WithPrefixes(c) {
+  var (x, xs) := xAndXs;
+  ...
+}
+```
+
+Where the signature of `WithPrefixes` is something like:
+
+```dafny
+function method WithPrefixes<T>(s: seq<T>): seq<(T, ghost seq<T>)> {
+  ...
+}
+```
+
+Similarly, a helper function can be provided to maintain a running index of the enumerated values:
+
+```dafny
+foreach xAndI <- WithIndexes(s) {
+  var (x, i) := xAndI;
+  ...
+}
+```
+
+## Sequence comprehensions
+
+A sequence comprehension has identical syntax to a set comprehension, except that it begins with
+`seq` rather than `set`. Therefore its syntax is:
+
+```
+SeqComprehension ::=
+  "seq" 
+  QuantifierDomain
+  [ :: Expression ]
+```
+
+Sequence comprehensions are likely to be easier to work with than `foreach` loops, and Dafny users should be
+encouraged to use them instead where possible. Many high-order operations on sequences can be expressed directly
+using sequence comprehensions:
+
+```dafny
+// Filtering to optional values
+var s: seq<Option<T>> := ...;
+var filtered: seq<T> := seq o <- s | o.Some? :: o.value;
+
+// Map and then flatten (a.k.a. "FlatMap")
+var c: seq<S> := ...;
+var f: S -> seq<T> := ...;
+var flatMapped: seq<T> := seq x <- c, y <- f(x) :: y;
+
+// Intersection of sequences
+var a := [4, 1, 3, 5];
+var b := [3, 1, 7];
+assert (seq x <- a | x in b) == [1, 3];
+assert (seq x <- b | x in a) == [3, 1];
+```
+
+Since sequence comprehensions are expressions rather than statements, they carry an additional restriction
+when compiled: their ordering must be fully deterministic. If `s` is a `set<int>`, for example,
```

Comment:
This is easy to ensure at the verification level. In fact, it difficult NOT to ensure it, because expressions in math are deterministic. This is the reason that the compiler also must make sure to do this consistently.

Reply:
I do think this is SUCH a common point of confusion, it needs to be taught and documented more prominently. We already have this problem with `:|` and it surprises even advance Dafny users.

Reply:
Still, I wonder whether it wouldn't be better to disallow `seq x <- someSet` in all contexts. What would be the use-case of doing this in a ghost context?

Reply:
You definitely need some efficient way of accessing the values in a set in sequence at runtime, but it might be enough to only support `foreach`. Supporting collections other than sequences as quantified variable domains is not until step 4 of the implementation plan, so I'm definitely happy to hold off until we feel there's a need for it.